File: libs/shapesearcher/src/main/kotlin/mono/shapesearcher/ShapeZoneAddressManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shapesearcher

import mono.graphics.bitmap.MonoBitmap
import mono.shape.shape.AbstractShape

/**
 * A model class to convert a shape to addresses which the shape have pixels on.
 * This class also cache the addresses belong to the shape along with its version.
 */
internal class ShapeZoneAddressManager(private val getBitmap: (AbstractShape) -> MonoBitmap?) {
    private val idToZoneAddressMap: MutableMap<String, VersionizedZoneAddresses> = mutableMapOf()

    fun getZoneAddresses(shape: AbstractShape): Set<ZoneAddress> {
        val cachedAddresses = getCachedAddresses(shape)
        if (cachedAddresses != null) {
            return cachedAddresses
        }

        val bitmap = getBitmap(shape)
        if (bitmap == null) {
            idToZoneAddressMap.remove(shape.id)
            return emptySet()
        }
        val position = shape.bound.position
        val addresses = mutableSetOf<ZoneAddress>()
        for (ir in bitmap.matrix.indices) {
            bitmap.matrix[ir].forEachIndex { ic, _, _ ->
                val row = ir + position.row
                val col = ic + position.column
                val address = ZoneAddressFactory.toAddress(row, col)
                addresses.add(address)
            }
        }
        val versionizedZoneAddresses = VersionizedZoneAddresses(shape.versionCode, addresses)
        idToZoneAddressMap[shape.id] = versionizedZoneAddresses
        return versionizedZoneAddresses.addresses
    }

    private fun getCachedAddresses(shape: AbstractShape): Set<ZoneAddress>? =
        idToZoneAddressMap[shape.id]?.takeIf { it.version == shape.versionCode }?.addresses

    private class VersionizedZoneAddresses(val version: Int, val addresses: Set<ZoneAddress>)
}

File: monosketch-svelte/src/lib/mono/shape/searcher/shape-zone-address-manager.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { MonoBitmap } from "$mono/monobitmap/bitmap/monobitmap";
import type { AbstractShape } from '$mono/shape/shape/abstract-shape';
import { ZoneAddressFactory, type ZoneAddress } from './zone-address';

interface VersionizedZoneAddresses {
    version: number;
    addresses: Set<ZoneAddress>;
}

/**
 * A class to convert a shape to addresses which the shape have pixels on.
 * This class also cache the addresses belong to the shape along with its version.
 */
export class ShapeZoneAddressManager {
    private idToZoneAddressMap: Map<string, VersionizedZoneAddresses> = new Map();

    constructor(private getBitmap: (shape: AbstractShape) => MonoBitmap.Bitmap | null) {
    }

    getZoneAddresses(shape: AbstractShape): Set<ZoneAddress> {
        const cachedAddresses = this.getCachedAddresses(shape);
        if (cachedAddresses) {
            return cachedAddresses;
        }

        const bitmap = this.getBitmap(shape);
        if (!bitmap) {
            this.idToZoneAddressMap.delete(shape.id);
            return new Set();
        }

        const position = shape.bound.position;
        const addresses = new Set<ZoneAddress>();
        for (let ir = 0; ir < bitmap.matrix.length; ir++) {
            for (let {index} of bitmap.matrix[ir].asSequence()) {
                const row = ir + position.row;
                const col = index + position.column;
                const address = ZoneAddressFactory.toAddress(row, col);
                addresses.add(address);
            }
        }

        const versionizedZoneAddresses = {version: shape.versionCode, addresses};
        this.idToZoneAddressMap.set(shape.id, versionizedZoneAddresses);
        return versionizedZoneAddresses.addresses;
    }

    private getCachedAddresses(shape: AbstractShape): Set<ZoneAddress> | null {
        const cached = this.idToZoneAddressMap.get(shape.id);
        return cached && cached.version === shape.versionCode ? cached.addresses : null;
    }
}
